File: KCD/HMMaskSelectView.m
//
//  HMMaskSelectView.m
//  KCD
//
//  Created by Hori,Masaki on 2014/09/05.
//  Copyright (c) 2014年 Hori,Masaki. All rights reserved.
//

#import "HMMaskSelectView.h"

const NSInteger kNumberOfMask = 4;
static NSRect maskRects[kNumberOfMask];

@implementation HMMaskSelectView

+ (void)initialize
{
	static dispatch_once_t onceToken;
	dispatch_once(&onceToken, ^{
		maskRects[0] = NSMakeRect(113, 455, 100, 18);
		maskRects[1] = NSMakeRect(203, 330, 200, 25);
		maskRects[2] = NSMakeRect(95, 378, 100, 18);
		maskRects[3] = NSMakeRect(60, 378, 100, 18);
	});
}

- (id)initWithFrame:(NSRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        // Initialization code here.
    }
    return self;
}

- (void)drawRect:(NSRect)dirtyRect
{
    NSGraphicsContext *context = [NSGraphicsContext currentContext];
	[context saveGraphicsState];
	[context setShouldAntialias:NO];
	
	[[NSColor redColor] set];
	CGFloat dashSeed[] = {3.0, 3.0};
	for(int i = 0; i < kNumberOfMask; i++) {
		NSBezierPath *path = [NSBezierPath bezierPathWithRect:maskRects[i]];
		[path setLineDash:dashSeed count:2 phase:0];
		[path stroke];
	}
	
	[context restoreGraphicsState];
	
	NSBezierPath *path = [NSBezierPath bezierPath];
	for(int i = 0; i < 800; i += 50) {
		[path moveToPoint:NSMakePoint(i, 0)];
		[path lineToPoint:NSMakePoint(i, 480)];
	}
	for(int j = 0; j < 480; j += 50) {
		[path moveToPoint:NSMakePoint(0, j)];
		[path lineToPoint:NSMakePoint(800, j)];
	}
	[[NSColor whiteColor] set];
	[path stroke];
}

- (void)mouseUp:(NSEvent *)event
{
	NSPoint mouse = [event locationInWindow];
	mouse = [self convertPoint:mouse fromView:nil];
	
	for(NSInteger i = 0; i < kNumberOfMask; i++ ) {
		if(NSMouseInRect(mouse, maskRects[i], self.isFlipped)) {
			NSLog(@"Click on Mask view.");
			break;
		}
	}
}

@end

File: KCD/HMMaskSelectView.swift
//
//  HMMaskSelectView.swift
//  KCD
//
//  Created by Hori,Masaki on 2014/12/30.
//  Copyright (c) 2014年 Hori,Masaki. All rights reserved.
//

import Cocoa

class HMMaskSelectView: NSView
{
	override init(frame: NSRect) {
		super.init(frame: frame)
		masks = makeMask(frame: frame)
	}
	required init?(coder: NSCoder) {
	    super.init(coder: coder)
		masks = makeMask(frame: frame)
	}
	
	let masks: [HMMaskInformation]!
	
	private func makeMask(frame frameRect: NSRect) -> [HMMaskInformation] {
		let originalSize: NSSize = NSSize(width: 800, height: 480)
		
		var masks: [HMMaskInformation] = []
		let widthRatio = frameRect.size.width / originalSize.width
		let heightRatio = frameRect.size.height / originalSize.height
		
		let maskRects: [NSRect] = [
			NSRect(x: 113, y: 455, width: 100, height: 18),
			NSRect(x: 203, y: 330, width: 200, height: 25),
			NSRect(x: 95, y: 378, width: 100, height: 18),
			NSRect(x: 60, y: 378, width: 100, height: 18)
		]
		
		for i in 0..<maskRects.count {
			let info = HMMaskInformation()
			info.maskRect = NSRect(
				x: maskRects[i].origin.x * widthRatio,
				y: maskRects[i].origin.y * heightRatio,
				width: maskRects[i].size.width * widthRatio,
				height: maskRects[i].size.height * heightRatio)
			if i == 3 {
				info.borderColor = NSColor(calibratedRed: 0.00, green: 0.011, blue: 1.000, alpha: 1.000)
			} else {
				info.borderColor = NSColor.redColor()
			}
			masks += [info]
		}
		return masks
	}
	
    override func drawRect(dirtyRect: NSRect) {
		let context = NSGraphicsContext.currentContext()
		context?.saveGraphicsState()
		context?.shouldAntialias = false
		
		let dashSeed:[CGFloat] = [3.0, 3.0]
		for info in masks {
			let path = NSBezierPath(rect: info.maskRect)
			if info.enable {
				info.maskColor.set()
				path.fill()
			}
			path.setLineDash(dashSeed, count: dashSeed.count, phase: 0)
			info.borderColor?.set()
			path.stroke()
			path.setLineDash(dashSeed, count: dashSeed.count, phase: 3.0)
			NSColor.lightGrayColor().set()
			path.stroke()
		}
		context?.restoreGraphicsState()
    }
	
	override func mouseUp(theEvent: NSEvent) {
		var mouse = theEvent.locationInWindow
		mouse = convertPoint(mouse, fromView: nil)
		for info in masks.reverse() {
			if NSMouseInRect(mouse, info.maskRect, flipped) {
				info.enable = !info.enable
				setNeedsDisplayInRect(NSInsetRect(info.maskRect, -5, -5))
				break
			}
		}
	}
	
	@IBAction func disableAllMasks(sender: AnyObject?) {
		for info in masks {
			info.enable = false
		}
		needsDisplay = true
	}
}
